### PR TITLE
[14.0] shopfloor: change search location action

### PR DESCRIPTION
**Pull request by TDu**
Change the search to first look on the barcode field and then on
location name.

### CODE REVIEW

File: shopfloor/actions/search.py [hunk 1]
```diff
@@ -14,13 +14,14 @@ class SearchAction(Component):
...
-    def location_from_scan(self, barcode):
+    def location_from_scan(self, barcode, limit=1):
         model = self.env["stock.location"]
         if not barcode:
             return model.browse()
-        return model.search(
-            ["|", ("barcode", "=", barcode), ("name", "=", barcode)], limit=1
-        )
+        res = model.search([("barcode", "=", barcode)], limit=limit)
+        if len(res) < limit:
```
Comment (simahawk) [inline shopfloor/actions/search.py:24]: I don't get why you need to play w/ limit :thinking:

File: shopfloor/actions/search.py [hunk 2]
```diff
@@ -14,13 +14,14 @@ class SearchAction(Component):
...
+            res += model.search([("name", "=", barcode)], limit=(limit - len(res)))
```
Comment (jbaudoux) [inline shopfloor/actions/search.py]: I think with += you could return twice the  same location
```suggestion
            res |= model.search([("name", "=", barcode)], limit=(limit - len(res)))
```

Comment (TDu) [reply shopfloor/actions/search.py]: I don't know about having multiple record in a set ? But better use the one listed in the docs https://www.odoo.com/documentation/14.0/developer/reference/addons/orm.html#operations.
Changed

File: shopfloor/actions/search.py (hunk 1 above)
Comment (TDu) [reply shopfloor/actions/search.py:24]: This was a requirement, @jbaudoux should know the underlying reason.

**Review by jbaudoux — APPROVED** [review]

File: shopfloor/actions/search.py (hunk 2 above)
Comment (simahawk) [reply shopfloor/actions/search.py]: Indeed, the docs talk about using `set` operators on recordsets :wink: 
```
>>> rset1
res.partner(14, 26)
>>> rset2
res.partner(14, 26, 33, 27)
>>> rset1 | rset2
res.partner(14, 26, 33, 27)
>>> rset1 + rset2
res.partner(14, 26, 14, 26, 33, 27)
```
and if you use `+=`
```
>>> rset3
res.partner(14, 26)
>>> rset3 += rset2
>>> rset3
res.partner(14, 26, 14, 26, 33, 27)
```
while if you use `|=`
```
>>> rset3
res.partner(14, 26)
>>> rset3 |= rset2
>>> rset3
res.partner(14, 26, 33, 27)
```

File: shopfloor/actions/search.py (hunk 1 above)
Comment (simahawk) [reply shopfloor/actions/search.py:24]: @jbaudoux ?
I guess the idea is: if we have a limit of 10 and we find only 3 results, we do another search by barcode to find the rest.
Having a comment on top would be nice.

Comment (jbaudoux) [reply shopfloor/actions/search.py:24]: The concern is that the search is not scoped to scenario/menu context. So to have a proper match, you must ensure location uniqueness which is not a standard odoo requirement.
Ideally we should be able to pass a domain & limit in the search. Having the limit is a first good improvement, you can then apply scope on the returned recordset

Comment (sebalix) [reply shopfloor/actions/search.py:24]: Having a comment on top is not only nice, it's a requirement I would say :)

Comment (TDu) [reply shopfloor/actions/search.py:24]: Comment added

**Review by sebalix — APPROVED** [review]

**Pull request merged**